### PR TITLE
remove global metadata cache, refactor custom_metadata API

### LINKED ISSUE #90: remove global metadata cache (OBJ_METADATA)
EDIT: I rescoped this issue, see https://github.com/JuliaData/Arrow.jl/issues/90#issuecomment-873094100

I noticed that metadata is stored in a global `IdDict` - would it make sense to provide an `unsetmetadata!(x)` (or use a sentinel e.g. `setmetadata!(x, nothing)`) that calls `delete!(OBJ_METADATA, x)`?

I could see a memory-leaky scenario where a e.g. a long running service writes a bunch of Arrow objects and attaches a small amount of metadata to each one and eventually OOMs or something.

### PR DESCRIPTION
**Pull request by jrevels**
essentially implements the approach described here https://github.com/JuliaData/Arrow.jl/issues/90#issuecomment-914870675

~I haven't run tests fully locally yet so may still have some lingering bugs.~ They pass 😎 

I also haven't been super performance-conscious here so we should be careful to make sure regressions aren't being introduced. 

Since we're doing a major bump for this anyway, I wonder whether it's a good time to fully delete some of the lingering deprecations here.

closes #90, #211

### CODE REVIEW

**Comment by jrevels** [comment]
Flipping out of draft as tests pass and docs are updated (unless I missed anything).

Before merging, should we decide on:

> Since we're doing a major bump for this anyway, I wonder whether it's a good time to fully delete some of the lingering deprecations here.

We also could do this in a separate PR before tagging

File: src/write.jl [hunk 1]
```diff
@@ -229,26 +198,28 @@ end
...
-        newcol = toarrowvector(col, i, dictencodings, dictencodingdeltas; compression=compress, largelists=largelists, denseunions=denseunions, dictencode=dictencode, dictencodenested=dictencodenested, maxdepth=maxdepth)
+        oldcolmeta = getmetadata(col)
+        newcolmeta = isnothing(colmeta) ? oldcolmeta : get(colmeta, nm, oldcolmeta)
+        newcol = toarrowvector(col, i, dictencodings, dictencodingdeltas, newcolmeta; compression=compress, largelists=largelists, denseunions=denseunions, dictencode=dictencode, dictencodenested=dictencodenested, maxdepth=maxdepth)
         newtypes[i] = eltype(newcol)
         newcols[i] = newcol
     end
     minlen, maxlen = isempty(newcols) ? (0, 0) : extrema(length, newcols)
     minlen == maxlen || throw(ArgumentError("columns with unequal lengths detected: $minlen < $maxlen"))
+    meta = isnothing(meta) ? nothing : toidict(String(k) => String(v) for (k, v) in meta)
```
Comment (ericphanson) [inline src/write.jl:222]:
```suggestion
    meta = _normalizemeta(meta)
```

**Review by ericphanson — COMMENTED** [review]
exciting! Didn't know about `Base.ImmutableDict`, cool.

File: src/utils.jl [hunk 2]
```diff
@@ -207,3 +207,13 @@ function tobuffer(data; kwargs...)
...
+
+toidict(x::Base.ImmutableDict) = x
+
+function toidict(pairs)
+    dict = Base.ImmutableDict(first(pairs))
+    for pair in Iterators.drop(pairs, 1)
+        dict = Base.ImmutableDict(dict, pair)
+    end
+    return dict
+end
```
Comment (quinnj) [inline src/utils.jl:219]:
```suggestion
toiddict(pairs) = foldl(Base.ImmutableDict, pairs)
```

**Review by quinnj — APPROVED** [review]
LGTM; thanks for taking this on! I left one suggestion to simplify `toiddict`.

File: src/utils.jl (hunk 2 above)
Comment (jrevels) [reply src/utils.jl:219]: actually had something like this originally but seemed like `foldl` was a bit slower than the more manual loop (didn't investigate why):

```jl
julia> pairs = [randstring(30) => randstring(30) for _ in 1:15];

julia> f(x) = foldl(Base.ImmutableDict, x)
f (generic function with 1 method)

julia> @benchmark f($pairs)
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.442 μs … 179.375 μs  ┊ GC (min … max): 0.00% … 98.91%
 Time  (median):     1.479 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.541 μs ±   3.071 μs  ┊ GC (mean ± σ):  3.44% ±  1.71%

       ▃ ▇ █ ▅ ▃ ▅ ▅ ▃ ▄ ▄ ▁
  ▂▃▃▇▁█▁█▁█▁█▁█▁█▁█▁█▁█▁█▁█▁█▁▇▁▇▁▆▁▅▁▅▁▄▁▄▁▄▁▃▁▃▁▂▁▂▁▂▁▂▁▂▂ ▄
  1.44 μs         Histogram: frequency by time        1.57 μs <

 Memory estimate: 1.66 KiB, allocs estimate: 60.

julia> @benchmark toidict($pairs)
BenchmarkTools.Trial: 10000 samples with 964 evaluations.
 Range (min … max):   90.076 ns …   1.652 μs  ┊ GC (min … max):  0.00% … 93.47%
 Time  (median):     102.524 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):   117.682 ns ± 152.728 ns  ┊ GC (mean ± σ):  13.49% ±  9.63%

  █▂                                                            ▁
  ██▄▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅ █
  90.1 ns       Histogram: log(frequency) by time       1.57 μs <
```

Maybe the magnitude here doesn't matter that much in reality though...you'd have to have small arrow messages and lots of metadata, which doesn't seem like a likely/good case to optimize for...

so let's go with the simpler `foldl` as you suggest and can optimize later if it ever matters

File: src/utils.jl [hunk 3]
```diff
@@ -207,3 +207,7 @@ function tobuffer(data; kwargs...)
     seekstart(io)
     return io
 end
+
+toidict(x::Base.ImmutableDict) = x
+
+toiddict(pairs) = foldl(Base.ImmutableDict, pairs)
```
Comment (jrevels) [inline src/utils.jl:213]:
```suggestion
toidict(pairs) = foldl(Base.ImmutableDict, pairs)
```

fix typo

**Comment by jrevels** [comment]
From Slack convo: 

I'll merge this PR and can open a follow-up later to delete the deprecated regions before we actually tag the v2.0 release.

**Comment by jrevels** [comment]
Hmm. Looks like Julia 1.3 tests now fail because the `foldl` implementation of `toidict` relies on `Base.ImmutableDict(::Pair, ::Pair)` working, which I guess isn't defined on Julia 1.3

and the Julia 1.6 tests fail because `foldl` doesn't apply the given `op` on unary inputs (which is expected):  

```jl
julia> foldl(Base.ImmutableDict, ("a" => "b",))
"a" => "b"
```

I think I'll just revert back to the original more verbose implementation

**Pull request merged**